### PR TITLE
fix(query): floor percentage minimum_should_match (BUG-403)

### LINKED ISSUE #403: fix(query): minimum_should_match percentage rounds up (ceil) instead of down (floor), over-restricting boolean queries
## Summary

`resolve_minimum_should_match` uses `f32::ceil()` to convert a percentage-based `minimum_should_match` into a required-term count. The correct rounding direction is `floor()` (round down), which the project already uses for analogous percentage-to-count conversions elsewhere (e.g. histogram bucketing at `aggs/mod.rs:4272` explicitly uses `floor()` to "match Elasticsearch semantics"). The `ceil()` rounding silently demands more matching terms than the user requested, excluding valid documents from results.

## Affected code

- `searchlite-core/src/query/planner.rs:1098-1100` — the `ceil()` call:
  ```rust
  let raw = (percent / 100.0) * term_count as f32;
  // 0% explicitly allows zero required matches; callers opt into this.
  raw.ceil() as usize
  ```

## Reproduction / evidence

Given a 3-term query with `minimum_should_match: "75%"`:

| Rounding | Computation         | Required terms | Behaviour                     |
|----------|---------------------|----------------|-------------------------------|
| `floor`  | `floor(0.75 × 3) = floor(2.25) = 2` | 2 of 3 | Docs matching 2+ terms returned (expected) |
| `ceil`   | `ceil(0.75 × 3)  = ceil(2.25)  = 3`  | 3 of 3 | Only docs matching ALL 3 terms returned     |

Additional examples where the two diverge:

- `"50%"` of 3 terms: `floor` → 1 required, `ceil` → 2 required
- `"34%"` of 3 terms: `floor` → 1 required, `ceil` → 2 required
- `"60%"` of 5 terms: `floor` → 3 required, `ceil` → 3 required (same)
- `"25%"` of 3 terms: `floor` → 0 required, `ceil` → 1 required

The impact is that any non-integer percentage-of-terms product that falls between two integers causes the engine to demand one more matching term than the user expected, silently dropping documents that should have been returned.

## Expected behavior

`floor()` rounding: `"75%"` of 3 terms requires 2 matches (not 3). This is the documented Elasticsearch behaviour ("The number computed from the percentage is rounded down and used as the minimum") and matches the project's own precedent in histogram bucketing (`aggs/mod.rs:4272-4278`).

## Actual behavior

`ceil()` rounding: `"75%"` of 3 terms requires 3 matches. Documents matching exactly 2 of 3 terms are silently excluded.

## Root cause (hypothesis)

The `ceil()` was likely chosen to handle the `0%` edge case (the comment at line 1099 calls this out). However, `floor(0.0) == 0` as well, so `floor()` handles 0% identically — `ceil()` is not needed for that case.

## Suggested fix

Replace `raw.ceil()` with `raw.floor()` (or cast via `raw as usize`, which truncates toward zero — equivalent to `floor()` for non-negative values):

```rust
let raw = (percent / 100.0) * term_count as f32;
raw.floor() as usize
```

Both `floor(0.0) = 0` and `floor(exact_integer) = exact_integer`, so all whole-number percentages (including 0% and 100%) produce the same results as before.

## Environment

- Commit: 1dc1582
- Rust MSRV in repo: 1.88.0
- Found by: scheduled Claude Code bug hunt

### PR DESCRIPTION
**Pull request by davidkelley**
## Summary

Closes #403.

`resolve_minimum_should_match` converts a percentage-based `minimum_should_match` (e.g. `"75%"`) into a required-term count. It used `f32::ceil()`, which silently demanded **one more** matching term than the user requested for any non-integer product — over-restricting boolean queries and dropping docs that should have been returned.

Switch to `floor()`, matching:
- Elasticsearch's documented behaviour (*"The number computed from the percentage is rounded down and used as the minimum"*).
- The project's own precedent for percentage-to-count conversion in histogram bucketing (`searchlite-core/src/query/aggs/mod.rs:603` uses `.floor() as usize`).

### Impact (before / after)

| Spec        | Terms | Before (`ceil`) | After (`floor`) |
|-------------|-------|-----------------|-----------------|
| `"75%"`     | 3     | 3               | **2**           |
| `"50%"`     | 3     | 2               | **1**           |
| `"34%"`     | 3     | 2               | **1**           |
| `"25%"`     | 3     | 1               | **0**           |
| `"60%"`     | 5     | 3               | 3 (unchanged)   |
| `"100%"`    | 4     | 4               | 4 (unchanged)   |
| `"0%"`      | 3     | 0               | 0 (unchanged)   |

Integer products (`60%` of 5, `100%` of 4) are identical under both roundings, so callers relying on whole-percentage math are unaffected. The `0%`-allows-zero-matches edge case (called out in the original comment) also survives because `floor(0.0) == 0`.

## Changes

- `searchlite-core/src/query/planner.rs` — `raw.ceil() as usize` → `raw.floor() as usize` in `resolve_minimum_should_match`, with an updated comment explaining the Elasticsearch-compat rationale.
- Added regression tests `resolve_minimum_should_match_percentage_rounds_down` and `resolve_minimum_should_match_percentage_caps_at_term_count` exercising every divergent case above plus the integer-product controls.

## Test plan

- [x] `cargo test -p searchlite-core --lib resolve_minimum_should_match` — 2/2 pass.
- [x] `cargo test -p searchlite-core --lib` — 506/506 pass (504 previous + 2 new).
- [x] `cargo clippy -p searchlite-core --all-targets -- -D warnings` — clean.
- [x] `cargo fmt --all --check` — clean.
- [x] Negative check: reverting the `floor()` back to `ceil()` fails `resolve_minimum_should_match_percentage_rounds_down` at the first case (`"75%"` of 3 → got 3, expected 2), confirming the test genuinely pins the semantic.

### CODE REVIEW

**Comment by davidkelley** [comment]
@codex review

File: searchlite-core/src/query/planner.rs [hunk 1]
```diff
@@ -1096,8 +1096,10 @@ fn resolve_minimum_should_match(
         bail!("minimum_should_match percentage must be between 0 and 100");
       }
       let raw = (percent / 100.0) * term_count as f32;
-      // 0% explicitly allows zero required matches; callers opt into this.
-      raw.ceil() as usize
+      // BUG-403: round down to match Elasticsearch `minimum_should_match`
+      // percentage semantics (e.g. 75% of 3 terms = 2, not 3). `floor(0.0)`
+      // also preserves the 0%-allows-zero-matches behaviour.
+      raw.floor() as usize
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/src/query/planner.rs]: Good catch — confirmed locally that `(13.0_f32 / 100.0) * 900.0` evaluates to `116.999992…`, floor = 116, dropping the mathematically correct 117. Fixed in f3f1dd6: whole-number percentages now compute as `(whole * term_count) / 100` in integer arithmetic (exact by construction), and fractional percentages compute as `percent * term_count / 100` in `f64` (reordered multiply-before-divide, so the intermediate stays exact for integer products). Added `("13%", 900, 117)`, `("70%", 10, 7)`, `("7%", 1000, 70)` to the regression table to pin it.

[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/src/query/planner.rs]: Addressed in f3f1dd6. Switched the percentage path to `f64` and — more importantly — changed the evaluation order to `percent * term_count / 100` so whole-number percentages give exact integer products (no `6.9999995`-style drift). For whole-number percentages the code now uses integer arithmetic (`(whole * term_count) / 100`), which is exact by construction. Added your suggested `("70%", 10, 7)` case, plus `("13%", 900, 117)` (the Codex-flagged sibling) and `("7%", 1000, 70)` to guard against regressions, plus fractional-percent coverage (`"12.5%"` of 8, `"33.3%"` of 3).

File: searchlite-core/src/query/planner.rs [hunk 2]
```diff
@@ -1641,6 +1643,48 @@ mod tests {
...
+        "{pct} of {term_count} terms: expected {expected}, got {got:?}"
+      );
+    }
+  }
+
+  #[test]
+  fn resolve_minimum_should_match_percentage_caps_at_term_count() {
+    // Values above 100% are rejected by the range check, but the final
+    // `.min(term_count)` still guards against any future drift.
+    let spec = Some(MinimumShouldMatch::Percentage("100%".to_string()));
```
[the remark this answers is not in the document]
Comment (davidkelley) [reply searchlite-core/src/query/planner.rs]: You are right that the previous test was misleading — with the `0..=100` range check, a percentage can never exceed `term_count`, so `.min(term_count)` is a no-op on that path. Replaced the test in f3f1dd6 with `resolve_minimum_should_match_value_caps_at_term_count`, which uses `MinimumShouldMatch::Value(10)` against a 4-term query so the cap actually fires. Also added float-rounding edge cases (`"12.5%"` of 8, `"33.3%"` of 3) to the rounds-down test alongside the integer-product regression cases that motivated your first comment.

**Comment by davidkelley** [comment]
@codex review

**Comment by davidkelley** [comment]
Rebased onto latest main (no content changes, just picking up #404). CI re-running.

@codex review
